libraries/chain/webassembly/eos-vm-oc/executor.cpp

```diff
@@ -0,0 +1,251 @@
+  boost::hana::index_if(intrinsic_table, ::boost::hana::equal.to(BOOST_HANA_STRING("eosvmoc_internal.grow_memory"))).value()
+);
+
+//This is effectively overriding the eosio_exit intrinsic in wasm_interface
+static void eosio_exit(int32_t code) {
+   EOSVMOC_MEMORY_PTR_cb_ptr;
+   siglongjmp(*cb_ptr->jmp, EOSVMOC_EXIT_CLEAN_EXIT);
+   __builtin_unreachable();
+}
+static intrinsic eosio_exit_intrinsic("env.eosio_exit", IR::FunctionType::get(IR::ResultType::none,{IR::ValueType::i32}), (void*)&eosio_exit,
+  boost::hana::index_if(intrinsic_table, ::boost::hana::equal.to(BOOST_HANA_STRING("env.eosio_exit"))).value()
+);
+
+static void throw_internal_exception(const std::string& s) {
+   EOSVMOC_MEMORY_PTR_cb_ptr;
+   *cb_ptr->eptr = std::make_exception_ptr(wasm_execution_error(FC_LOG_MESSAGE(error, s)));
+   siglongjmp(*cb_ptr->jmp, EOSVMOC_EXIT_EXCEPTION);
+   __builtin_unreachable();
+}
+
+#define DEFINE_EOSVMOC_TRAP_INTRINSIC(module,name) \
+	void name(); \
+	static intrinsic name##Function EOSVMOC_INTRINSIC_INIT_PRIORITY(#module "." #name,IR::FunctionType::get(),(void*)&name, \
+     boost::hana::index_if(intrinsic_table, ::boost::hana::equal.to(BOOST_HANA_STRING(#module "." #name))).value() \
+   ); \
+	void name()
+
+DEFINE_EOSVMOC_TRAP_INTRINSIC(eosvmoc_internal,depth_assert) {
+   throw_internal_exception("Exceeded call depth maximum");
+}
+
+DEFINE_EOSVMOC_TRAP_INTRINSIC(eosvmoc_internal,div0_or_overflow) {
+   throw_internal_exception("Division by 0 or integer overflow trapped");
+}
+
+DEFINE_EOSVMOC_TRAP_INTRINSIC(eosvmoc_internal,indirect_call_mismatch) {
+   throw_internal_exception("Indirect call function type mismatch");
+}
+
+DEFINE_EOSVMOC_TRAP_INTRINSIC(eosvmoc_internal,indirect_call_oob) {
+   throw_internal_exception("Indirect call index out of bounds");
+}
+
+DEFINE_EOSVMOC_TRAP_INTRINSIC(eosvmoc_internal,unreachable) {
+   throw_internal_exception("Unreachable reached");
+}
+
+executor::executor(const code_cache_base& cc) {
+   //XXX
+   static bool once_is_enough;
+   if(!once_is_enough) {
+      printf("current_linear_memory_pages %li\n", -memory::cb_offset + offsetof(eosio::chain::eosvmoc::control_block, current_linear_memory_pages));
+      printf("full_linear_memory_start %li\n", -memory::cb_offset + offsetof(eosio::chain::eosvmoc::control_block, full_linear_memory_start));
+      printf("cb_offset %lu\n", memory::cb_offset);
+      printf("remaining call depth: %li\n", -memory::cb_offset + offsetof(eosio::chain::eosvmoc::control_block, current_call_depth_remaining));
+      printf("first intrinsic: %lu\n", memory::first_intrinsic_offset);
+      printf("current code offset: %li\n", -memory::cb_offset + offsetof(eosio::chain::eosvmoc::control_block, running_code_base));
+      once_is_enough = true;
+   }
+
+   //if we're the first executor created, go setup the signal handling. For now we'll just leave this attached forever
+   if(std::lock_guard g(inited_signal_mutex); inited_signal == false) {
```

Comment:
local static initialization is thread-safe.  You can just create a dummy variable instead of using a mutex+bool like this.